safe/load.go

```diff
@@ -3,117 +3,89 @@ package safe
-	}
-
-	safeDTO := safeDTO{}
-	if err := json.Unmarshal(data, &safeDTO); err == nil { // nolint: vetshadow
-		// Unmarshalling did succeed -> Safe uses new format
-		// Do nothing
-	} else {
-		safeDTO.Ciphertext = data
-		// If we don't have a config, use OpenPGP for backwards-compatibility
-		defaultOpenPGPConfig := crypto.NewDefaultConfigWithType(crypto.ConfigTypeOpenPGP)
-		safeDTO.Config = &defaultOpenPGPConfig
-	}
-
-	// We first try to decrypt a safe using its own configuration
-	s.crypto, err = crypto.New(safeDTO.Config)
-	if err != nil {
-		return nil, err
 	}
 
-	upgradeCrypto := encryptionClient
-	plaintext, err := s.crypto.Decrypt(safeDTO.Ciphertext, password)
+	var configurationHasChanged bool
+	plaintext, err := decryptSafeDTO(ciphertext, password, nil)
 	if err != nil {
-		// Wasn't able to decrypt the safe with its own configuration.
-		// Normally this shouldn't happen, never. As a fallback, we try
-		// to decrypt using the user-provided config. If this fails, the
-		// password is _definitely_ incorrect. If however decryption works
-		// now, we need to upgrade the safe to use the user-provided config.
-		// Therefore, store current crypto for possible upgrade.
-		upgradeCrypto = s.crypto
-		s.crypto = encryptionClient
-		plaintext, err = s.crypto.Decrypt(safeDTO.Ciphertext, password)
+		// Failed to decrypt the safe with its own configuration,
+		// fall back to the user-provided config.
+		plaintext, err = decryptSafeDTO(ciphertext, password, encryptionClient)
 		if err != nil {
+			// The password is definitely incorrect
 			return nil, err
+		} else {
+			configurationHasChanged = true
 		}
 	}
 
 	var tmp Safe
-
 	if err := json.Unmarshal(plaintext, &tmp); err != nil { // nolint: vetshadow
 		return nil, errors.ErrSafeCorrupt
 	}
-
 	if tmp.Config != nil && tmp.Config.Version != "" {
-		incompatible, err := versionIncompatible(tmp.Config.Version, config.Version) // nolint: vetshadow
-		if err != nil {
+		if err := tmp.IsVersionCompatible(config.Version); err != nil { // nolint: vetshadow
 			return nil, err
-		} else if incompatible {
-			return nil, fmt.Errorf("Safe is using a non-backwards compatible version. Please upgrade pick. pick version: %s, safe version: %s", config.Version, tmp.Config.Version)
 		}
 	}
 
-	s.Accounts = tmp.Accounts
-	s.Notes = newNotesManager(&s)
+	var notes map[string]note
 	if tmp.Notes != nil {
-		s.Notes.Notes = tmp.Notes.Notes
+		notes = tmp.Notes.Notes
 	}
 
-	// Default 'ModifiedOn' to 'CreatedOn' if it is empty
-	for i, acc := range s.Accounts {
-		if acc.ModifiedOn == 0 {
-			acc.ModifiedOn = acc.CreatedOn
-			s.Accounts[i] = acc
-		}
+	safe, err := NewSafe(password, backend, encryptionClient, config, tmp.Accounts, notes)
+	if err != nil {
+		return nil, err
 	}
 
-	// We need to compare the default / user-provided config with the safe config.
-	// If they differ -> Upgrade safe
-	if !reflect.DeepEqual(*safeDTO.Config, s.Config.Encryption) {
-		fmt.Println("Upgrading safe")
```

Comment:
Where did this go?

Reply:
Good catch- I thought I was handling that case in the logic using `decryptSafeDTO` but after another look I am not.

I'll get this case covered.

Reply:
I was referring to the `fmt.Println` call. Not using `reflect.DeepEqual` should be fine, right?

Reply:
No, it's still required. Take this scenario:

- Safe (stored) is encrypted with OpenPGP
- `~/.pick/config.toml` mandates to use AES

1. Safe is decrypted using its own config, succeeds
2. Safe is not upgraded :o

We need unit tests for this.

Reply:
We could remove the `configurationHasChanged` flag and instead only rely on the deep equal check.